### ISSUE TITLE
[Linux] Build instructions for Linux?

### ISSUE THREAD

**Issue by MegaBrutal** [ValveSoftware/halflife#3133]
How am I supposed to build the server and client .so files (`hl.so,` `client.so`) for Linux?

I see there are makefiles in the `dlls` and `linux` subdirectories, but they don't work right off the bat; guess I'm supposed to make a special build environment for it. For example, `linux/Makefile` assumes I have a compiler at `/valve/bin/g++-4.6` (yeah, I could symlink `gcc`, but then I noticed that the other Makefile seems to be newer, so I left it be); while `dlls/Makefile` seems to assume a much older version of GCC, as modern versions don't know the `-m486` argument. I realized that the argument has changed to `-march=486` and `-m32` is also required to explicitly tell that I want to build 32-bit code (which is again a surprise, I thought Half-Life is on 64-bit already), but then it started to miss various `.h` files.

Sure I'm not the only one who wants to build the DLLs (SO files) for Linux, so if anyone succeeded so far, please shed some light!

As a personal note, I have customized DLLs from like 2007 for a mod I thought to make but never saw the light. Meanwhile I have completely moved to Linux and only have Windows on VMs and computers I don't care about. Thought it would be fun to resurrect the project and play my own mod on my native operating system. Also it's surprising and wonderful to see that people still make mods for Half-Life 1 even to this day, but sadly they rarely compile their customized DLLs for Linux – actually, I'm yet to see a fan made mod that supports Linux (and has custom server/client libraries to begin with). Maybe it would be nice to start a trend. :)

**Comment by tschumann** [comment]
Use the one in linux/ as the one in dlls/ is outdated and won't build
client.so
I updated the path to just use system gcc and that works.
Half-Life may have had a 64 bit server but the game is still 32 bit.
Which header files are missing? You'll need to install a package with 32
bit headers on newer Linux versions (I can't remember which but it'll
depend on your distro anyway).

**Comment by solidi** [comment]
@MegaBrutal - see my open pull request below. This successfully builds .dll, .so, and .dylib libs and publishes them directly on github on every commit. I hope this helps.
 https://github.com/ValveSoftware/halflife/pull/3113

**Comment by gabuch2** [comment]
I have exactly the same problem as @[MegaBrutal](https://github.com/MegaBrutal), although I'm using this repo codebase instead of Solokiller's.

I had to adjust some defines to allow compiling on modern Linux tools but after the successful build the game refuses to load the compiled binaries, sending the exact same error as described.

Has anyone had any success building this?

**Comment by MegaBrutal** [comment]
@gabuch2 Yes, but you're definitely better off with Solokiller's (now SamVanheer) fork, because it's updated to compile on modern systems and I think it also has various other fixes.

My problem was that I had HLDS (Half-Life Dedicated Server) installed and it ships outdated versions of the standard C/C++ libraries which overrides the system default libraries your game SO files are (supposedly) built against. So your SO files require newer standard C/C++ libraries than the ones presented for them.

See this issue for details:
https://github.com/SamVanheer/halflife-updated/issues/33

Either uninstall HLDS if not absolutely necessary (you can always install it with SteamCMD, I personally use a Docker container); or you can safely delete the problematic libraries as it doesn't do any harm (older applications seem to work with newer libraries, just not the other way around), but then updates may put them back.

Aside from HLDS, you may have built your libraries on a newer system than you're trying to use them.

**Comment by f1am3d** [comment]
I have similar question but platform independent.
What C standard should be used? 
Compilator throws errors, like `error C2668: 'abs': ambiguous call to overloaded function` which does mean that some specific version of C++ should be used.
But repository has no information about this.

**Comment by tschumann** [comment]
I think the abs error is just bit-rot - updated it to fabs in the problematic places and that should fix the error.
